### PR TITLE
Fix supertable bugs

### PR DESCRIPTION
**Pull request by SaptakS**
Refs: https://github.com/freedomofpress/freedom.press/issues/983

- Fixes the logic to make rowspan and colspan persist properly
- Adds a basic template and css file, that supports alignment, merge cells and coloring
- Adds templatetags to get rowspan and colspan from classnames, since wagtail table JS code is written in a way that only classname meta persists and it's difficult to override that functionality since it is called in multiple occurences. Added a template tag (similar to wagtail's cell_classname template_tag) that helps to parse and get rowspan and colspan from the classnames
- ~Adds merge cells array in the settings for handsontable to persist~ Uses jQuery to get rowspan and colspan from persistent classnames and add the necessary attributes via DOM manipulation for admin side

image

This is how the above table looks with the default HTML/CSS provided with supertable:
image

### CODE REVIEW

**Comment by chigby** [comment]
I installed this branch on freedom.press locally (using https://github.com/freedomofpress/wagtail-supertable/zipball/284f75844996e7a3ab67e48a64b5e494b21d0f20 in requirements.in) and I am getting a lot of this error: `Uncaught TypeError: $cell.attr(...) is undefined`.

Stack trace:
image

Not sure if there's something else I need to be doing? It is happening when I make a new StreamPage and insert a table block in the body, then edit a value in the cell (not trying to merge cells).

**Comment by SaptakS** [comment]
Interesting. I tested with securethenews actually and it worked. I will test with freedom.press also and see what's the issue.

**Comment by SaptakS** [comment]
Okay. When there are not tables initiated or no merge cell, it was throwing this error because I needed to check for `$cell.length` for jquery object returned and not just `$cell`. That should be fixed in the above commit.

**Comment by SaptakS** [comment]
Also, just to note, for testing this with freedom.press, since freedom.press overwrites the table.html used for rendering a table in the frontend, this PR itself won't fix rendering of the merge cells in the page generated by wagtail. That you can test by using it in some other repo, which doesn't overwrite the table.html.

If testing with freedom.press, the only thing to check is once you merge cell, and save draft or publish, and then go to edit the page again, it should still show the cells merged.

**Comment by chigby** [comment]
Ok, here's the latest on what I'm doing to test this. It seems like a lot of steps but each one is fairly small.

1. Remove wagtail-supertable line from requirements.in
2. Run `make compile-pip-dependencies`
3. Add this line to docker-compose.yaml under `services.django.volumes`. Use the path to your local repository for supertable.
```
- /home/chigby/fpf/wagtail-supertable:/opt/wagtail-supertable
```
4. In the freedom.press repo, edit `devops/docker/django-start.sh`. Add this line to the top of the function `django_start()`. It has to come before `./manage.py migrate`:
```
pip install -e /opt/wagtail-supertable
```
5. Run `docker-compose up --build` and the wagtail supertable should be installed in editable mode from the local git repo, meaning editing the repo (either locally or updating it from github) should be found and used "live" by the running website. It seems to be working all right! Though it is possibly not easier than copying the directory manually.

Note: in trying to get the JS to update live as well, I got a webpack error on the command `npm run start`. I think the `start` script in package.json needs updating for webpack 5. I ran into a similar situation on freedom.press & related sites when migrating to webpack 5.

**Comment by SaptakS** [comment]
That might be possible. I was only doing `npm run build` since when we install from git, the built JS file is already there

**Comment by chigby** [comment]
While playing around with the merging of cells I found some behavior I'm not sure I understand. The table gets a little confused and one of the cells goes away (the bottom right corner) after I start entering text if I've merged multiple cells. I'm not sure how to explain it exactly so here's an animation of what I'm doing to reproduce the problem:
c

File: wagtailsupertable/templatetags/table_span_tags.py
Comment (chigby) [inline wagtailsupertable/templatetags/table_span_tags.py]: We are doing a lot of work here (splitting, checking if a substring is present, replacing text) that could also be done by the regular expression directly. This is maybe a personal preference thing, but another way of writing the regex to combine all these operations into one might be:

```python
r = re.compile(r"\bcolspan-(?P<colspan>\d+)\b")
```

I'm using a named group here and the [\b word boundary matcher](https://docs.python.org/3/howto/regex.html#more-metacharacters). This regexp lets us operate directly on the entire css class string and select part of the match we care about (the colspan amount) by name:

```python
match = r.search(cell_class)
if match:
    colspan = match.group('colspan')
```

The regular expression itself is a little more involved by in my opinion the rest of the code is cleaner. But I leave it up to you if you like this approach.

File: wagtailsupertable/templatetags/table_span_tags.py
Comment (chigby) [inline wagtailsupertable/templatetags/table_span_tags.py]: Regular expression match objects return strings by default, so I think this comparison ought to be with `'1'`.

**Review by chigby — APPROVED** [review]
The latest updates to this definitely fixed the strangeness I was seeing, so that's great! I left a few comments here and there about the code, which I think are not blockers but might be things you want to change or leave alone, depending. Anyway, this issue was pretty tricky to unravel, so nice work clearing it up!

File: wagtailsupertable/templatetags/table_span_tags.py
Comment (SaptakS) [reply wagtailsupertable/templatetags/table_span_tags.py]: I do like the word boundary matcher idea. Might need to just add some comments so I remember what it is doing when I come back to it.

**Pull request merged**
